### PR TITLE
Canary deployment

### PR DESCRIPTION
**Pull request by eedorenko**

### CODE REVIEW

File: docs/canary_ab_deployment.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+"New Model B"
+"New Model A"
+"New Model B"
+"New Model B"
+"New Model A"
+"New Model A"
+"New Model A"
+```
+
+Despite what blue/green weights are configured now on the cluster, you can perform ***A/B testing*** and send requests directly to either blue or green images:
```
Comment (dtzar) [inline docs/canary_ab_deployment.md]: Explain breifly how this works (i.e. header matching).

File: code/scoring/scoreA.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+def init():
```
[the remark this answers is not in the document]
Comment (dtzar) [reply code/scoring/scoreA.py]: A better approach would be to parameterize the score.py script with "data" as the default response with ability to specify "Model A" or "Model B".

File: charts/abtest-model/values.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+appname: model
+
+deployment:
+  name: model-green  
+  bluegreen: green
+  container:
+    name: model  
+    port: 5001
+  image:
+    name: amlabtest74c95585.azurecr.io/abtestimg
```
[the remark this answers is not in the document]
Comment (dtzar) [reply charts/abtest-model/values.yaml]: It's actually OK as long as it gets overrode during helm upgrade (which isn't clear how the tag & image repo gets updated in the template alone).  I'd suggest changing to something like `overrideme.azurecr.io`

File: .pipelines/azdo-abtest-pipeline.yml [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+          steps:
+          - script: |
+              IMAGE_LOCATION="$(cat $(Pipeline.Workspace)/image_location/image_location.txt)"
+              echo "##vso[task.setvariable variable=IMAGE_LOCATION]$IMAGE_LOCATION"
+            displayName: 'Get Image Location'
+          - template: azdo-helm-upgrade.yml
+            parameters:
+              chartPath: '$(Pipeline.Workspace)/allcharts/abtest-model'
+              releaseName: $(greenReleaseName) 
+              overrideValues: 'deployment.name=$(greenReleaseName),deployment.bluegreen=green,initialDeployment=true,deployment.image.name=$(IMAGE_LOCATION)'
```
Comment (dtzar) [inline .pipelines/azdo-abtest-pipeline.yml]: `intialDeployment=true` ok every time?

File: charts/abtest-model/values.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    tag: 2
```
Comment (dtzar) [inline charts/abtest-model/values.yaml]: It seems tag isn't used in your chart.  It's best practice to keep tag vs use just one string for the whole repo:tag.

File: ml_service/util/create_scoring_image.py [hunk 6]
```diff
@@ -29,7 +29,7 @@
 os.chdir("./code/scoring")
 
 image_config = ContainerImage.image_configuration(
-    execution_script="score.py",
+    execution_script=e.score_script,
```
Comment (dtzar) [inline ml_service/util/create_scoring_image.py]: Need to add this VAR to the pipeline variables template now with default value of `score.py`

File: .pipelines/azdo-abtest-pipeline.yml (hunk 4 above)
Comment (eedorenko) [reply .pipelines/azdo-abtest-pipeline.yml]: Good catch. it's garbage. Not used anymore

File: charts/abtest-model/values.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (eedorenko) [reply charts/abtest-model/values.yaml]: Removed

File: ml_service/util/create_scoring_image.py (hunk 6 above)
Comment (eedorenko) [reply ml_service/util/create_scoring_image.py]: It is in azdo-variables.yml

File: charts/abtest-model/values.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (dtzar) [reply charts/abtest-model/values.yaml]: Still there and tag too.

File: code/scoring/scoreA.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eedorenko) [reply code/scoring/scoreA.py]: Added explanation

File: docs/canary_ab_deployment.md (hunk 1 above)
Comment (eedorenko) [reply docs/canary_ab_deployment.md]: Added explanation

File: charts/abtest-model/values.yaml (hunk 5 above)
Comment (eedorenko) [reply charts/abtest-model/values.yaml]: Removed

**Pull request merged**